### PR TITLE
build(dockerized_app): add dockerized app

### PR DESCRIPTION
**Pull request by AlexMomot-717**

### CODE REVIEW

File: .dockerignore
Comment (svntmr) [inline .dockerignore:1]: `.dockerignore` files work in the build context, is it the right location for this file?

File: README.Docker.md
Comment (svntmr) [inline README.Docker.md:1]: what's the point of this file? imo it's better to include docker instructions into project README

File: compose.yaml [hunk 1]
```diff
@@ -1,6 +1,11 @@
-version: "3.9"
+version: "3"
```
Comment (svntmr) [inline compose.yaml]: what was the reason for the compose version downgrade?

File: compose.yaml [hunk 2]
```diff
@@ -1,6 +1,11 @@
...
 
 services:
+  app:
+    container_name: "page_analyzer_app"
```
Comment (svntmr) [inline compose.yaml]: should it just be `app` or `backend`?

File: compose.yaml
Comment (svntmr) [inline compose.yaml:1]: let's keep the default name

File: compose.yaml [hunk 3]
```diff
@@ -11,7 +16,3 @@ services:
     env_file: docker-images/postgres/.env.default
     volumes:
       - postgres:/var/lib/postgres/data
-
-volumes:
-  postgres:
-    name: postgres-volume
```
Comment (svntmr) [inline compose.yaml]: what was the reason of removing this volume?

File: .env.example [hunk 4]
```diff
@@ -2,3 +2,6 @@ SECRET_KEY = "YOUR APP SECRET KEY GOES HERE"
 
 # database connection URL in format {provider}://{user}:{password}@{host}:{port}/{db}
 DATABASE_URL = "postgresql://janedoe:mypassword@localhost:5432/mydb"
+
+# database connection URL for tests
+DATABASE_URL_TEST = "postgresql://janedoe:mypassword@localhost:5432/testdb"
```
Comment (svntmr) [inline .env.example]: this environment variable is not necessary, we can keep `DATABASE_URL` and adjust it to "prod" and tests database URLs

File: tests/conftest.py
Comment (svntmr) [inline tests/conftest.py:1]: changes can be reverted as there is no point in having extra environment variable

File: docker-images/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,23 @@
+ARG PYTHON_VERSION=3.11.7
+FROM python:${PYTHON_VERSION}-slim as base
```
Comment (svntmr) [inline docker-images/Dockerfile]: you can remove `ARG` if you're not going to build this docker image with different versions of python

File: docker-images/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+# Prevents Python from writing pyc files.
+ENV PYTHONDONTWRITEBYTECODE=1
+
+# Keeps Python from buffering stdout and stderr to avoid situations where
+# the application crashes without emitting any logs due to buffering.
+ENV PYTHONUNBUFFERED=1
+
+WORKDIR .. .
```
Comment (svntmr) [inline docker-images/Dockerfile]: `WORKDIR`'s usually have a name

File: docker-images/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+# Download dependencies from pyproject.toml
+#RUN pip install
+
+# Copy the source code into the container.
+COPY .. .
```
Comment (svntmr) [inline docker-images/Dockerfile]: at this point we need to:
- copy poetry files (pyproject and lock)
- copy project files (only the package itself)

then we need to tell poetry not to create virtual environments and install project dependencies (like in github actions)

File: docker-images/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+# Expose the port that the application listens on.
+EXPOSE 8000
+
+# Run the application.
+CMD poetry run gunicorn -w 5 -b 0.0.0.0:8000 page_analyzer:app
```
Comment (svntmr) [inline docker-images/Dockerfile]: `CMD` written as list of strings (aka "exec form") is the best practice - https://docs.docker.com/reference/dockerfile/#cmd

**Review by svntmr — CHANGES_REQUESTED** [review]

File: docker-images/.dockerignore
Comment (svntmr) [inline docker-images/.dockerignore:1]: let's move backend docker files into `backend` folder, so the full path is going to be `docker-images/backend/...`

File: docker-images/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+RUN pip install poetry
+
+# Skipping virtualenv creation
+RUN poetry config virtualenvs.create false
+
+# Install dependencies
+RUN poetry install
+
+# configure the container to run in an executed manner
+ENTRYPOINT ["poetry", "run"]
```
Comment (svntmr) [inline docker-images/Dockerfile]: should we try this?
```suggestion
ENTRYPOINT ["gunicorn", "-b", "0.0.0.0:8000", "page_analyzer:app"]
```

File: tests/conftest.py [hunk 10]
```diff
@@ -29,7 +29,7 @@ def test_database() -> Generator[None, None, None]:
     """
     os.environ[
         "DATABASE_URL"
-    ] = "postgresql://page-analyzer:page-analyzer@localhost:5432/page-analyzer"
+    ] = "postgresql://alexandra:Alisa+717@localhost:5432/test_db"
```
Comment (svntmr) [inline tests/conftest.py]: is it a second DB you run in postgres container? can't find its code anywhere

**Review by svntmr — CHANGES_REQUESTED** [review]
wdyt, should we concentrate on adding a second test database first?

File: Makefile [hunk 11]
```diff
@@ -35,5 +35,17 @@ refresh-postgres-data:
...
 
+build-services:
+	@docker-compose build --no-cache
+
 up:
-	docker-compose up
+	@docker-compose up -d
+
+down:
+	@docker-compose down --volumes
```
Comment (svntmr) [inline Makefile]: volumes are usually persisted, are you sure you want to delete database data each time?

File: README.md [hunk 12]
```diff
@@ -3,3 +3,13 @@
 [![pre-commit hooks](https://github.com/AlexMomot-717/python-project-83/actions/workflows/pre-commit.yml/badge.svg)](https://github.com/AlexMomot-717/python-project-83/actions)
 [![Maintainability](https://api.codeclimate.com/v1/badges/7ad5b8e6d1a22a4fdf34/maintainability)](https://codeclimate.com/github/AlexMomot-717/python-project-83/maintainability)
 [![Test Coverage](https://api.codeclimate.com/v1/badges/7ad5b8e6d1a22a4fdf34/test_coverage)](https://codeclimate.com/github/AlexMomot-717/python-project-83/test_coverage)
+
+### Building and running your application
```
Comment (svntmr) [inline README.md]: usually, this section is called `setup` or `how to run the application`

File: README.md [hunk 13]
```diff
@@ -3,3 +3,13 @@
...
+
+When you're ready, start application by running:
+
+```
+docker compose up --build
+```
```
Comment (svntmr) [inline README.md]: let's use make here (as we have targets already) to explain how to start the application

part with "when you're ready ..." is not needed

File: README.md [hunk 14]
```diff
@@ -3,3 +3,13 @@
...
+
+Your application will be available at http://0.0.0.0:8000.
```
Comment (svntmr) [inline README.md]: is it the reader's app? should "your" be used here?

**Review by svntmr — CHANGES_REQUESTED** [review]
final polishing

File: Makefile (hunk 11 above)
Comment (AlexMomot-717) [reply Makefile]: `docker-compose down` also removes database data

Comment (svntmr) [reply Makefile]: it shouldn't, down only stops and removes containers and networks by default
it won't do anything with volumes, that's why you're providing an extra flag here

Comment (svntmr) [reply Makefile]: context: https://docs.docker.com/reference/cli/docker/compose/down/

File: README.md [hunk 15]
```diff
@@ -3,3 +3,33 @@
...
 [![Test Coverage](https://api.codeclimate.com/v1/badges/7ad5b8e6d1a22a4fdf34/test_coverage)](https://codeclimate.com/github/AlexMomot-717/python-project-83/test_coverage)
+
+### Setup:
+
+The application will be available at http://0.0.0.0:8000.
+Run the application using following commands:
+
+1. Build or rebuild application services images:
+   ```
+   make bild-services
```
Comment (svntmr) [inline README.md]: did you mean build?
```suggestion
   make build-services
```

File: README.md [hunk 16]
```diff
@@ -3,3 +3,33 @@
 [![pre-commit hooks](https://github.com/AlexMomot-717/python-project-83/actions/workflows/pre-commit.yml/badge.svg)](https://github.com/AlexMomot-717/python-project-83/actions)
 [![Maintainability](https://api.codeclimate.com/v1/badges/7ad5b8e6d1a22a4fdf34/maintainability)](https://codeclimate.com/github/AlexMomot-717/python-project-83/maintainability)
 [![Test Coverage](https://api.codeclimate.com/v1/badges/7ad5b8e6d1a22a4fdf34/test_coverage)](https://codeclimate.com/github/AlexMomot-717/python-project-83/test_coverage)
+
+### Setup:
+
```
Comment (svntmr) [inline README.md:14]: usually, it goes from the steps that need to be taken and then to the expected result

here you start from the expected result and then list all the steps

File: README.md [hunk 17]
```diff
@@ -3,3 +3,33 @@
...
+The application will be available at http://0.0.0.0:8000.
+Run the application using following commands:
+
```
Comment (svntmr) [inline README.md]: wdyt, is it necessary to know all the commands you have prepared if they just want to start the application? `make up` should be enough

**Review by svntmr — CHANGES_REQUESTED** [review]

File: README.md (hunk 16 above)
Comment (svntmr) [reply README.md:14]: setup block should've stayed at the same place, this comment was about it's content :)

**Review by svntmr — APPROVED** [review]
please, address my latest comment and we're good to go

**Review by svntmr — APPROVED** [review]

**Pull request merged**